smcontroller/smcontroller.go

```diff
@@ -139,410 +122,354 @@ func New(
-		controller.readyWG.Add(1)
+	controller.grpcServer = grpc.NewServer(opts...)
 
-		go controller.connectClient(ctx, smConfig, secureOpt)
-	}
+	pb.RegisterSMServiceServer(controller.grpcServer, controller)
+
+	go func() {
+		if err := controller.startServer(cfg.SMController.SMServerURL); err != nil {
+			log.Errorf("Can't start SM controller server: %v", err)
+		}
+	}()
 
 	return controller, nil
 }
 
 // Close closes SM controller.
-func (controller *Controller) Close() (err error) {
+func (controller *Controller) Close() error {
 	log.Debug("Close SM controller")
 
-	controller.cancelFunction()
-	controller.clientsWG.Wait()
-
-	controller.Lock()
-	defer controller.Unlock()
-
-	for _, client := range controller.clients {
-		client.close()
-	}
+	controller.stopServer()
 
 	return nil
 }
 
-// GetServicesStatus returns SM all existing services status.
-func (controller *Controller) GetServicesStatus() (servicesStatus []unitstatushandler.ServiceStatus, err error) {
-	controller.waitAndLock()
-	clients := controller.clients
-	controller.Unlock()
-
-	for _, client := range clients {
-		clientServices, err := client.getServicesStatus()
-		if err != nil {
-			return nil, aoserrors.Wrap(err)
-		}
+// GetNodeConfiguration gets node static configuration.
+func (controller *Controller) GetNodeConfiguration(nodeID string) (cfg launcher.NodeConfiguration, err error) {
+	controller.Lock()
+	defer controller.Unlock()
 
-		servicesStatus = append(servicesStatus, clientServices...)
+	handler, err := controller.getNodeHandlerByID(nodeID)
+	if err != nil {
+		return cfg, aoserrors.Wrap(err)
 	}
 
-	return servicesStatus, nil
+	return handler.config, nil
 }
 
-// GetUsersStatus returns SM users status.
-func (controller *Controller) GetLayersStatus() (layersStatus []unitstatushandler.LayerStatus, err error) {
-	controller.waitAndLock()
-	clients := controller.clients
-	controller.Unlock()
-
-	for _, client := range clients {
-		clientLayers, err := client.getLayersStatus()
-		if err != nil {
-			return nil, aoserrors.Wrap(err)
-		}
+// GetUnitConfigStatus gets unit configuration status fot he node.
+func (controller *Controller) GetUnitConfigStatus(nodeID string) (string, error) {
+	controller.Lock()
+	defer controller.Unlock()
 
-		layersStatus = append(layersStatus, clientLayers...)
+	handler, err := controller.getNodeHandlerByID(nodeID)
+	if err != nil {
+		return "", aoserrors.Wrap(err)
 	}
 
-	return layersStatus, nil
+	return handler.getUnitConfigState()
 }
 
-// CheckBoardConfig checks board config.
-func (controller *Controller) CheckBoardConfig(boardConfig aostypes.BoardConfig) error {
-	controller.waitAndLock()
-	clients := controller.clients
-	controller.Unlock()
-
-	// we do not support multiple SM right now, check the same config for all SM's
+// CheckUnitConfig checks unit config for the node.
+func (controller *Controller) CheckUnitConfig(nodeCfg aostypes.NodeConfig, vendorVersion string) error {
+	controller.Lock()
```

Comment:
Looks like you don't need to lock all the API, lock getNodeHandlerByID inside would be enough. Once you get handler, it is safe to perform multithreaded operation.